File: common/src/main/java/juuxel/adorn/lib/registry/Registered.java
package juuxel.adorn.lib.registry;

import net.minecraft.registry.RegistryKey;

import java.util.function.Supplier;

@FunctionalInterface
public interface Registered<T> extends Supplier<T> {
    interface WithKey<R, T extends R> extends Registered<T> {
        RegistryKey<R> key();
    }
}

File: common/src/main/kotlin/juuxel/adorn/lib/registry/Registered.kt
package juuxel.adorn.lib.registry

import java.util.function.Supplier
import kotlin.properties.ReadOnlyProperty
import kotlin.reflect.KProperty

fun interface Registered<out T> : Supplier<@UnsafeVariance T>, ReadOnlyProperty<Any?, T> {
    override fun get(): T

    override fun getValue(thisRef: Any?, property: KProperty<*>): T = get()
}
